routing/pathfind.go

```diff
@@ -623,6 +665,21 @@ func findPath(tx *bolt.Tx, graph *channeldb.ChannelGraph,
 			// any more up to date information concerning the
 			// bandwidth of this edge.
 			edgeBandwidth, ok := bandwidthHints[edgeInfo.ChannelID]
+
+			// If the destination is source, we need to use the capacity
+			// of the distant edge for the incoming path.
+			// Example: Local_ballance = 9 and distant = 1 then BandwidthHints
+			// uses 9 instead of 1, resulting in errors.  So we must subtract
+			// bandwidthHints from total capacty, 10-9 =1
+			// Checks nodeHeap to make sure we're not reversing capacity while
+			// exiting source.
+			if nodeHeap.Len() != 0 && (edgeInfo.NodeKey1Bytes == targetNode.PubKeyBytes ||
+				edgeInfo.NodeKey2Bytes == targetNode.PubKeyBytes) {
+				edgeBandwidth = lnwire.NewMSatFromSatoshis(
+                                        edgeInfo.Capacity,
+				) - edgeBandwidth
+			}
+
```

Comment:
I still think it is difficult to verify the correctness of your modifications. You add:
- Two code blocks for loop detection (normally not necessary for Dijkstra) 
- Rely on integer overflow behaviour
- Using distance 0, infinity and infinity-1 for the source node to signal states
- Flipping edge bandwidth

I still think adding a second virtual self node (earlier comment) results in higher quality code, because the core of the algorithm remains unchanged and also is in line with descriptions of the Dijkstra algorithm.

Reply:
Wouldn't the 'virtual self' have to have all the same channel ID's as the 'self' for proper route creation?
Then you run into "How do you avoid trying to route through the virtual node when not sending to self".  Would you just create/delete the virtual node on demand?  How would that be affected by a node with massive number of channels?  Or frequent balancing (my end goal is self-balancing channels that are triggered when funds go in/out of a channel causing it to be unbalanced)?

One of the blocks for 'loop detection' doesn't just check for loops, it's useful by always ignoring edges to nodes you just came from.  The other I had to add, because, somehow, a loop was created between two nodes directly connected to me.  [All three nodes are connected by zero fee channels, if that makes a difference.]

I'll do some research on adding the 'virtual self' to the channel graph.  I could probably manipulate the data flow to ignore all paths -> vSelf if destination != vSelf.  Then, whenever a channel is changed, it'd have to affect the vSelf is well, if is left on the map.

Reply:
Yes, it will have the same channel ids, but that is also what you want isn't it? Use any of that set to leave self and return via another one from the set. 

If you are not sending to self, you just don't add the virtual self. Nothing different happens at all compared to the current algorithm. So indeed, only when routing to self, you create the virtual self. 

Massive number of channels is no problem, it is just an additional edge for every channel, O(n). 

I don't see why frequent balancing would be different with either of the algorithms.

Dijkstra doesn't work with cycles. Three nodes with zero fee channels form a cycle. But normally the time lock delta adds some weight to the edge too, so it shouldn't happen. I think that case needs further investigation on what happened exactly. Checking for loops shouldn't be necessary.

I would not try to add virtual self to the global channel graph, that sounds like a scary change. I'd just do it locally inside findPath using additionalEdges. It should be quite simple and clean.

Reply:
I'm stuck:
How do I reference the channel policy of the node on the other end of the channel?
Can't assign the virtual edges the policy *from* self, it needs to be a copy of the policy *to* self.

                        edgeInfo *channeldb.ChannelEdgeInfo,
                        outEdge, _ *channeldb.ChannelEdgePolicy) error {

                        // If the remote balance is > amount add to vSelf edge to additional
                        // edges, else ignore.  The edge policy must be from the point of view
                        // of the non-self nodes.
                        edgeBandwidth, ok := bandwidthHints[edgeInfo.ChannelID]

                        if lnwire.NewMSatFromSatoshis(edgeInfo.Capacity) - edgeBandwidth > amt{

                                tovSelfEdge := &channeldb.ChannelEdgePolicy{
                                        Node:                                        vSelf,
                                        ChannelID:                                 outEdge.ChannelID,
          ***                          FeeBaseMSat:                            outEdge.node.FeeBaseMSat,
          ***                          FeeProportionalMillionths:       .FeeProportionalMillionths,
           ***                         TimeLockDelta:                         .TimeLockDelta,
                                }

Reply:
If you run LightningNode.ForEachChannel on your self node, it will return both the outgoing edge policies as well as the incoming edge policies of the remote node. Those incoming ones can be copied with self replaced by vself. Only when the search is done, the modified incoming edge policies need to be reverted to the original. Maybe by filling a lookup table when the vself edges are created.

Reply:
"If you run LightningNode.ForEachChannel on your self node," that's exactly what I'm doing.  I do not understand the syntax to reference the channel policy data of the remote node.   I just need the syntax. The policies of self never need to change, or be copied.  vSelf doesn't need outgoing channel policies as AdditionalEdges only needs the channel policy leading to vSelf.
Would "outEdge.ChannelID.outEdge.node.feeBaseMSat" be the FeeBaseMSat for the distant node policy for the channel back to self?
err := sourceNode.ForEachChannel(tx, func(tx *bolt.Tx,
                        edgeInfo *channeldb.ChannelEdgeInfo,
                        outEdge, _ *channeldb.ChannelEdgePolicy) error {

                        // If the remote balance is > amount add to vSelf edge to additional
                        // edges, else ignore.  The edge policy must be from the point of view
                        // of the non-self nodes.
                        edgeBandwidth, ok := bandwidthHints[edgeInfo.ChannelID]

                        if lnwire.NewMSatFromSatoshis(edgeInfo.Capacity) - edgeBandwidth > amt{

                                tovSelfEdge := &channeldb.ChannelEdgePolicy{
                                        Node:                      vSelf,
                                        ChannelID:                 outEdge.ChannelID,
                                        FeeBaseMSat:               outEdge.ChannelID.outEdge.node.FeeBaseMSat,
                                        FeeProportionalMillionths: outEdge.ChannelID.outEdge.node.FeeProportionalMillionths,
                                        TimeLockDelta:             outEdge.ChannelID.outEdge.node.TimeLockDelta,
                                }

                        additionalEdges := map[Vertex][]*channeldb.ChannelEdgePolicy{
                                NewVertex(outEdge.node.pubKeyBtes): {tempEdge},
                        }
                }

Reply:
You have to replace that `_` in the call to `ForEachChannel` to `inEdge`

Reply:
I'm almost done, just need to make it more efficient.  Considering I undid *everything* I posted in this PR, and started over, how do I post with a single commit, just the changes in the new file.  [Using GitHub only, I strongly dislike command line.]
Or do I just delete this PR, and make an entirely new one?

Reply:
Yes, I can imagine, the approach with virtual self is quite different. Hope you agree that it is better and/or more readable and that I didn't overlook something.

You can remove your old branch, rename your new branch to the old one and force push that to github. But I don't know how to do it without cmd line. It is probably useful to invest a little into command line knowledge anyway, as the project likes to have a clean commit history which requires rebasing.